File: ios/Classes/TiktokOpenSdkAuthPlugin.swift
import Flutter
import UIKit
import TikTokOpenSDK // Ensure this is the correct import for the TikTok SDK

public class TiktokOpenSdkAuthPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    private var channel: FlutterMethodChannel?
    private var authApi: TikTokOpenSDKApplicationDelegate? // Replace with actual TikTok Auth SDK type
    private var activity: UIViewController?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "tiktok_open_sdk_auth", binaryMessenger: registrar.messenger())
        let instance = TiktokOpenSdkAuthPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register for lifecycle events to manage activity state
        if let appDelegate = UIApplication.shared.delegate as? FlutterAppDelegate {
            appDelegate.add(instance)
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
                self.activity = rootViewController
                self.authApi = TikTokOpenSDKApplicationDelegate() // Initialize your TikTok SDK instance here
                result(true)
            } else {
                result(FlutterError(code: "NO_ACTIVITY", message: "TikTok Auth requires a foreground activity.", details: nil))
            }
        case "authorize":
            guard let authApi = self.authApi else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "TikTok Auth API is not initialized.", details: nil))
                return
            }
            let args = call.arguments as! [String: Any]
            let authMethod = args["authMethod"] as! Int
            let clientKey = args["clientKey"] as! String
            let scope = args["scope"] as! String
            let redirectUri = args["redirectUri"] as! String
            let codeVerifier = PKCEUtils.generateCodeVerifier()
            
            let authRequest = AuthRequest(
                clientKey: clientKey,
                scope: scope,
                redirectUri: redirectUri,
                codeVerifier: codeVerifier
            )
            
            authApi.authorize(authRequest, authMethod: AuthApi.AuthMethod(rawValue: authMethod)!) { response in
                if response.isSuccess {
                    result(true)
                } else {
                    result(FlutterError(code: "AUTH_FAILED", message: "Authorization failed.", details: response.errorDescription))
                }
            }
        case "getAuthResponseFromIntent":
            guard let authApi = self.authApi, let activity = self.activity else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "TikTok Auth API is not initialized.", details: nil))
                return
            }
            let args = call.arguments as! [String: Any]
            let redirectUri = args["redirectUri"] as! String
            let response = authApi.getAuthResponseFromIntent(activity, redirectUri: redirectUri)
            
            if response != nil {
                result(response)
            } else {
                result(FlutterError(code: "NO_RESPONSE", message: "No auth response found.", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // Handling lifecycle methods to manage the activity state
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
        return true
    }
    
    public func applicationDidBecomeActive(_ application: UIApplication) {
        // Handle activity becoming active if needed
    }
    
    public func applicationWillResignActive(_ application: UIApplication) {
        // Handle activity resigning active if needed
    }
    
    public func applicationDidEnterBackground(_ application: UIApplication) {
        // Handle entering background if needed
    }
    
    public func applicationWillEnterForeground(_ application: UIApplication) {
        // Handle entering foreground if needed
    }
}

File: android/src/main/kotlin/tiktok/tiktok_open_sdk_auth/TiktokOpenSdkAuthPlugin.kt
package tiktok.tiktok_open_sdk_auth

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.tiktok.open.sdk.auth.AuthApi
import com.tiktok.open.sdk.auth.AuthRequest
import com.tiktok.open.sdk.auth.AuthResponse
import com.tiktok.open.sdk.auth.utils.PKCEUtils

/** TiktokOpenSdkAuthPlugin */
class TiktokOpenSdkAuthPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel : MethodChannel
  private var activity: Activity? = null
  private var context: Context? = null
  private var authApi: AuthApi? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "tiktok_open_sdk_auth")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "init" -> {
        if (activity == null) {
          result.error("NO_ACTIVITY", "TikTok Auth requires a foreground activity.", null)
          return
        } 
        authApi = AuthApi(activity!!)
        result.success(true)
      }
      "authorize" -> {
        if (authApi == null) {
          result.error("NOT_INITIALIZED", "TikTok Auth API is not initialized.", null)
          return
        }

        val authMethod = call.argument<Int>("authMethod")!!
        val clientKey = call.argument<String>("clientKey")!!
        val scope = call.argument<String>("scope")!!
        val redirectUri = call.argument<String>("redirectUri")!!
        val callVerifier = PKCEUtils.generateCodeVerifier()

        val authRequest = AuthRequest(
            clientKey,
            scope,
            redirectUri,
            callVerifier
        )
        
        var response = authApi!!.authorize(authRequest, AuthApi.AuthMethod.values()[authMethod])
        result.success(true)
      }
      "getAuthResponseFromIntent" -> {
        if (authApi == null) {
          result.error("NOT_INITIALIZED", "TikTok Auth API is not initialized.", null)
          return
        }
        val redirectUri = call.argument<String>("redirectUri")!!
        val response = authApi!!.getAuthResponseFromIntent(activity!!.intent, redirectUri)

        result.success(true)
      }
      else -> {
        result.notImplemented()
      }
    }
  }
}
